# Project

---

## Scope & Ground Rules


### Part 1 - 50% | Scope

**it's composed for `5` small assignments with guidelines. The assignment scores **evenly distributed across all questions** (each question accounts for 10% of the final score). **


**Any question please contact to me via Slack or Email.**


---

**IMPORTANT NOTES to have in mind** 

**a) Code Readability is taken into account for the evaluation, so please make it simple, readable and explain your operations when necessary.**

**b) Make sure that the evaluater can re-run the notebook from the begining, i.e. before you delivery the assignment please go to the bar on top of your notebook -> `Kernel` -> `Restart & Run all`. Validate that all outputs are as you expect.**

----




# Part 1

## 1- Load Data

**1.1- Load **house_prices_final_project.csv** to a Pandas DataFrame. You can see in `data_description.txt` file the description of each column**

---

### Importação de bibliotecas

O seguinte script utiliza várias bibliotecas para preparação e análise de dados. Aqui está uma breve explicação de cada importação:

- `pandas` (`pd`): Utilizado para manipulação e análise de dados através da utilização de DataFrames.
- `KBinsDiscretizer` (de `sklearn.preprocessing`): Realiza discretização de dados contínuos em intervalos.
- `SimpleImputer` (de `sklearn.impute`): Preenche valores ausentes com estratégias como média, mediana ou valor constante.
- `OneHotEncoder` (de `sklearn.preprocessing`): Codifica variáveis categóricas em representações numéricas binárias (one-hot encoding).
- `MinMaxScaler` (de `sklearn.preprocessing`): Escala os valores para um intervalo específico (por exemplo, entre 0 e 1).
- `split_dataset` (de `pgds_mpp_utils`):  função personalizada para dividir um dataset em treino e teste.
- `score_approach` (de `pgds_mpp_utils`): função personalizada para treinar e avaliar modelo  e retornar a precisão .


In [917]:
import pandas as pd 
from sklearn.preprocessing import KBinsDiscretizer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import MinMaxScaler
from pgds_mpp_utils import split_dataset, score_approach

Código abaixo carrega os dados de um ficheiro CSV para um DataFrame do pandas:

In [918]:

df = pd.read_csv('data/house_prices_final_project.csv')

Apresentação dos primeiros 5 registos para confirmação de correta importação do passo anterior.

In [919]:
df.head(5)  

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

**1.2- Print dataset total number of `observations` and `variables`**

---

O bloco de código a seguir é utilizado para obter as dimensões do DataFrame e apresentar o número de linhas (observações) e colunas (variáveis):


In [920]:
observations, variables = df.shape

print(f"Número de observações: {observations} | Número de variáveis: {variables}")

Número de observações: 1460 | Número de variáveis: 81


---

### Please find below the subset of columns we are going to consider for the rest of the assignment

In [921]:
columns_list = ['FullBath',
                'TotRmsAbvGrd',
                'Fireplaces',
                'GarageYrBlt',
                'GarageCars',
                'GarageArea',
                'LotFrontage',
                'WoodDeckSF',
                'OpenPorchSF',
                'SaleType',
                'SaleCondition',
                'SalePrice']

**1.3- Create a new dataframe which is a subset of the origin dataframe based on the columns listed above.**

---

O bloco de código a seguir cria um subconjunto do DataFrame com colunas selecionadas e exibe as primeiras linhas para confirmar o resultado:

In [922]:
# Criar DataFrame com  subconjunto de colunas selecionadas
df_subset = df[columns_list]

# Exibir as primeiras linhas para confirmação de passo anterior
df_subset.head()

FullBath  TotRmsAbvGrd  Fireplaces  GarageYrBlt  GarageCars  GarageArea  \
0         2             8           0       2003.0           2         548   
1         2             6           1       1976.0           2         460   
2         2             6           1       2001.0           2         608   
3         1             7           1       1998.0           3         642   
4         2             9           1       2000.0           3         836   

   LotFrontage  WoodDeckSF  OpenPorchSF SaleType SaleCondition  SalePrice  
0         65.0           0           61       WD        Normal     208500  
1         80.0         298            0       WD        Normal     181500  
2         68.0           0           42       WD        Normal     223500  
3         60.0           0           35       WD       Abnorml     140000  
4         84.0         192           84       WD        Normal     250000

In [923]:
# Usar o info() consultar mem usage e da contagem de non null values de cada col
df_subset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   FullBath       1460 non-null   int64  
 1   TotRmsAbvGrd   1460 non-null   int64  
 2   Fireplaces     1460 non-null   int64  
 3   GarageYrBlt    1379 non-null   float64
 4   GarageCars     1460 non-null   int64  
 5   GarageArea     1460 non-null   int64  
 6   LotFrontage    1201 non-null   float64
 7   WoodDeckSF     1460 non-null   int64  
 8   OpenPorchSF    1460 non-null   int64  
 9   SaleType       1460 non-null   object 
 10  SaleCondition  1460 non-null   object 
 11  SalePrice      1460 non-null   int64  
dtypes: float64(2), int64(8), object(2)
memory usage: 137.0+ KB


## 2- Creating Labels

**2.1- Create the `target` column based on `SalePrice`. The split should be done using the median value to create 2 new buckets. `Min->Median` bucket should have assigned the value `0` while the other bucket (`Median->Max`) value should be `1`.**



**Note: you are free to decide the buckets boundaries**


---

O comando abaixo é utilizado para gerar estatísticas descritivas do DataFrame `df_subset` e transpor a saída para uma visualização mais clara:

In [924]:
df_subset.describe().T

count           mean           std      min       25%  \
FullBath      1460.0       1.565068      0.550916      0.0       1.0   
TotRmsAbvGrd  1460.0       6.517808      1.625393      2.0       5.0   
Fireplaces    1460.0       0.613014      0.644666      0.0       0.0   
GarageYrBlt   1379.0    1978.506164     24.689725   1900.0    1961.0   
GarageCars    1460.0       1.767123      0.747315      0.0       1.0   
GarageArea    1460.0     472.980137    213.804841      0.0     334.5   
LotFrontage   1201.0      70.049958     24.284752     21.0      59.0   
WoodDeckSF    1460.0      94.244521    125.338794      0.0       0.0   
OpenPorchSF   1460.0      46.660274     66.256028      0.0       0.0   
SalePrice     1460.0  180921.195890  79442.502883  34900.0  129975.0   

                   50%       75%       max  
FullBath           2.0       2.0       3.0  
TotRmsAbvGrd       6.0       7.0      14.0  
Fireplaces         1.0       1.0       3.0  
GarageYrBlt     1980.0    2002.0    2010.0  
GarageCars         2.0       2.0       4.0  
GarageArea       480.0     576.0    1418.0  
LotFrontage       69.0      80.0     313.0  
WoodDeckSF         0.0     168.0     857.0  
OpenPorchSF       25.0      68.0     547.0  
SalePrice     163000.0  214000.0  755000.0

O bloco de código a seguir utiliza o `KBinsDiscretizer` para categorizar a coluna `SalePrice` em dois grupos (bins) com base na estratégia de quantis:


In [925]:

# Criar o KBinsDiscretizer para 2 bins usando a estratégia quantile
discretizer = KBinsDiscretizer(n_bins=2, encode='ordinal', strategy='quantile')

# Ajustar e transformar os valores de SalePrice
df_subset['SalesPriceTarget'] = discretizer.fit_transform(df_subset[['SalePrice']]).astype(int).flatten()

print(df_subset['SalesPriceTarget'].value_counts())


SalesPriceTarget
1    732
0    728
Name: count, dtype: int64


C:\Users\PereiraRodrigues\AppData\Local\Temp\ipykernel_10840\2765499866.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_subset['SalesPriceTarget'] = discretizer.fit_transform(df_subset[['SalePrice']]).astype(int).flatten()


In [926]:
df_subset.sample(10)

FullBath  TotRmsAbvGrd  Fireplaces  GarageYrBlt  GarageCars  GarageArea  \
1271         2             7           1       1968.0           2         462   
937          2             8           1       2005.0           2         689   
800          2             8           0       1997.0           2         603   
88           1             6           0          NaN           0           0   
876          1             5           0       1963.0           2         648   
1289         2             8           1       2006.0           3         753   
567          2             7           0       2004.0           2         532   
1457         2             9           2       1941.0           1         252   
1380         1             5           0       1938.0           1         200   
264          1             4           0       1965.0           2         504   

      LotFrontage  WoodDeckSF  OpenPorchSF SaleType SaleCondition  SalePrice  \
1271          NaN           0            0       WD        Normal     185750   
937          75.0           0           48       WD        Normal     253000   
800          79.0         403          114       WD        Normal     200000   
88          105.0           0            0    ConLD       Abnorml      85000   
876          94.0           0            0       WD        Normal     132250   
1289         86.0         177           74      New       Partial     281000   
567          70.0           0            0       WD        Normal     214000   
1457         66.0           0           60       WD        Normal     266500   
1380         45.0           0            0       WD        Normal      58500   
264          30.0           0            0       WD        Normal      73000   

      SalesPriceTarget  
1271                 1  
937                  1  
800                  1  
88                   0  
876                  0  
1289                 1  
567                  1  
1457                 1  
1380                 0  
264                  0

## 3- Handling Missing Values

**3.1- List the amount of missing values per column**

---

O bloco de código seguinte identifica e lista o número de valores ausentes em cada coluna do DataFrame `df_subset`:

In [927]:
# Listar o número de valores ausentes por coluna
missing_values = df_subset.isnull().sum()
missing_values = missing_values[missing_values > 0].sort_values(ascending=False)
missing_values


LotFrontage    259
GarageYrBlt     81
dtype: int64

**3.2- Take care of the missing values in the column `LotFrontage`**

---

Como é possivel ver acima  na célua de estatísticas descritivas através do comando `describe()`, verifica-se que a mediana (69) está próxima à média (70.05), indicando uma distribuição aproximadamente simétrica, no entanto existem  valores altos como 313, sugerindo a possibilidade de outliers.
Abaixo é feito a verificação de outliers segundo o metodo intequartis



In [928]:
# Calculate IQR
Q1 = df_subset['LotFrontage'].quantile(0.25)
Q3 = df_subset['LotFrontage'].quantile(0.75)
IQR = Q3 - Q1

# Define the IQR boundaries
lower_bound_iqr = Q1 - 1.5 * IQR
upper_bound_iqr = Q3 + 1.5 * IQR

# Detect outliers using IQR
df_subset['LotFrontage_Outlier'] = df_subset['LotFrontage'].map(lambda x: 1 if x < lower_bound_iqr or x > upper_bound_iqr else 0)


# Count the number of outliers
outlier_count = df_subset['LotFrontage_Outlier'].sum()
outlier_info = {
    "Total Outliers": outlier_count,
    "Lower Bound": lower_bound_iqr,
    "Upper Bound": upper_bound_iqr
}

outlier_info

C:\Users\PereiraRodrigues\AppData\Local\Temp\ipykernel_10840\1046580130.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_subset['LotFrontage_Outlier'] = df_subset['LotFrontage'].map(lambda x: 1 if x < lower_bound_iqr or x > upper_bound_iqr else 0)


{'Total Outliers': np.int64(88),
 'Lower Bound': np.float64(27.5),
 'Upper Bound': np.float64(111.5)}

Verificando que existe cerca de 88 registos outliers, foi utilizado a técnica de imputação para preencher os valores ausentes na coluna `LotFrontage` com base na mediana que é mais robusta a outliers.

In [929]:
# Criando o imputer com a estratégia de mediana
imputer = SimpleImputer(strategy='median')

# Aplicando o imputer na coluna LotFrontage
df_subset[['LotFrontage']] = imputer.fit_transform(df_subset[['LotFrontage']])

# Verificando se há valores nulos após a imputação
missing_after_imputation = df_subset['LotFrontage'].isnull().sum()
print(f'Valores nulos após imputação: {missing_after_imputation}')

# Eliminar coluna de identificação de outlier, porque já não é preciso
df_subset.drop('LotFrontage_Outlier', axis=1)


Valores nulos após imputação: 0


C:\Users\PereiraRodrigues\AppData\Local\Temp\ipykernel_10840\1107123994.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_subset[['LotFrontage']] = imputer.fit_transform(df_subset[['LotFrontage']])


FullBath  TotRmsAbvGrd  Fireplaces  GarageYrBlt  GarageCars  GarageArea  \
0            2             8           0       2003.0           2         548   
1            2             6           1       1976.0           2         460   
2            2             6           1       2001.0           2         608   
3            1             7           1       1998.0           3         642   
4            2             9           1       2000.0           3         836   
...        ...           ...         ...          ...         ...         ...   
1455         2             7           1       1999.0           2         460   
1456         2             7           2       1978.0           2         500   
1457         2             9           2       1941.0           1         252   
1458         1             5           0       1950.0           1         240   
1459         1             6           0       1965.0           1         276   

      LotFrontage  WoodDeckSF  OpenPorchSF SaleType SaleCondition  SalePrice  \
0            65.0           0           61       WD        Normal     208500   
1            80.0         298            0       WD        Normal     181500   
2            68.0           0           42       WD        Normal     223500   
3            60.0           0           35       WD       Abnorml     140000   
4            84.0         192           84       WD        Normal     250000   
...           ...         ...          ...      ...           ...        ...   
1455         62.0           0           40       WD        Normal     175000   
1456         85.0         349            0       WD        Normal     210000   
1457         66.0           0           60       WD        Normal     266500   
1458         68.0         366            0       WD        Normal     142125   
1459         75.0         736           68       WD        Normal     147500   

      SalesPriceTarget  
0                    1  
1                    1  
2                    1  
3                    0  
4                    1  
...                ...  
1455                 1  
1456                 1  
1457                 1  
1458                 0  
1459                 0  

[1460 rows x 13 columns]

## 4- Handling Categorical Data

**4.1- Split categorical feature into a `df_categorical` dataframe**

---

In [930]:
df_subset

FullBath  TotRmsAbvGrd  Fireplaces  GarageYrBlt  GarageCars  GarageArea  \
0            2             8           0       2003.0           2         548   
1            2             6           1       1976.0           2         460   
2            2             6           1       2001.0           2         608   
3            1             7           1       1998.0           3         642   
4            2             9           1       2000.0           3         836   
...        ...           ...         ...          ...         ...         ...   
1455         2             7           1       1999.0           2         460   
1456         2             7           2       1978.0           2         500   
1457         2             9           2       1941.0           1         252   
1458         1             5           0       1950.0           1         240   
1459         1             6           0       1965.0           1         276   

      LotFrontage  WoodDeckSF  OpenPorchSF SaleType SaleCondition  SalePrice  \
0            65.0           0           61       WD        Normal     208500   
1            80.0         298            0       WD        Normal     181500   
2            68.0           0           42       WD        Normal     223500   
3            60.0           0           35       WD       Abnorml     140000   
4            84.0         192           84       WD        Normal     250000   
...           ...         ...          ...      ...           ...        ...   
1455         62.0           0           40       WD        Normal     175000   
1456         85.0         349            0       WD        Normal     210000   
1457         66.0           0           60       WD        Normal     266500   
1458         68.0         366            0       WD        Normal     142125   
1459         75.0         736           68       WD        Normal     147500   

      SalesPriceTarget  LotFrontage_Outlier  
0                    1                    0  
1                    1                    0  
2                    1                    0  
3                    0                    0  
4                    1                    0  
...                ...                  ...  
1455                 1                    0  
1456                 1                    0  
1457                 1                    0  
1458                 0                    0  
1459                 0                    0  

[1460 rows x 14 columns]

O código abaixo seleciona as colunas categóricas do DataFrame `df_subset` e cria um novo DataFrame apenas com essas colunas. Além disso, exibe informações sobre o DataFrame resultante:

In [931]:
# Criar DataFrame apenas com as colunas categóricas 
df_categorical = df_subset[df_subset.select_dtypes(include=['object']).columns.tolist()]

# Exibir as colunas categóricas selecionadas
print(df_categorical.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   SaleType       1460 non-null   object
 1   SaleCondition  1460 non-null   object
dtypes: object(2)
memory usage: 22.9+ KB
None


In [932]:
df_categorical.shape

(1460, 2)

O bloco de código abaixo converte colunas categóricas do tipo `object` para o tipo `category` no DataFrame `df_categorical` para otimização de utilização de memória e exibe informações sobre o resultado:

In [933]:
# Function to display the memory usage in MB for either a) a Pandas DataFrame or b) a Pandas Serie

def mem_usage(pandas_obj):
    if isinstance(pandas_obj,pd.DataFrame):
        usage_b = pandas_obj.memory_usage(deep=True).sum()
    else: # we assume if not a df it's a series
        usage_b = pandas_obj.memory_usage(deep=True)
    usage_mb = usage_b / 1024 ** 2 # convert bytes to megabytes
    return "{:03.4f} MB".format(usage_mb)

Convertendo colunas para o tipo `category` para melhorar a eficiência na utilização de memória e apresentação dos resultados

In [934]:
# Criação de variaveis categorizadas, e apresentação de tipo de dados 

df_categorical['SaleTypeCat'] = df_categorical['SaleType'].astype('category')
df_categorical['SaleConditionCat'] = df_categorical['SaleCondition'].astype('category')
df_categorical.info()

# Apresentar utilização de memória 

print('*Informação Uso de Memória de Variaveis*')
print(f"UsoSaleType: {mem_usage(df_categorical['SaleType'])}")
print(f"SaleTypeCat: {mem_usage(df_categorical['SaleTypeCat'])}")
print(f"SaleCondition: {mem_usage(df_categorical['SaleCondition'])}")
print(f"SaleConditionCat: {mem_usage(df_categorical['SaleConditionCat'])}")

df_categorical.drop(columns=['SaleType', 'SaleCondition'],inplace=True)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   SaleType          1460 non-null   object  
 1   SaleCondition     1460 non-null   object  
 2   SaleTypeCat       1460 non-null   category
 3   SaleConditionCat  1460 non-null   category
dtypes: category(2), object(2)
memory usage: 26.4+ KB
*Informação Uso de Memória de Variaveis*
UsoSaleType: 0.0714 MB
SaleTypeCat: 0.0023 MB
SaleCondition: 0.0769 MB
SaleConditionCat: 0.0020 MB


C:\Users\PereiraRodrigues\AppData\Local\Temp\ipykernel_10840\2488068249.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_categorical['SaleTypeCat'] = df_categorical['SaleType'].astype('category')
C:\Users\PereiraRodrigues\AppData\Local\Temp\ipykernel_10840\2488068249.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_categorical['SaleConditionCat'] = df_categorical['SaleCondition'].astype('category')
C:\Users\PereiraRodrigues\AppData\Local\Temp\ipykernel_10840\2488068249.py:15: SettingWithCopyWarn

**4.2- Apply OHE to `SaleType`**

In [935]:

# Inicializar o OneHotEncoder
ohe = OneHotEncoder(sparse_output=False)  

# Ajustar o encoder e transformar os dados
salestype_ohe = ohe.fit_transform(df_categorical[['SaleTypeCat']])

# Criar um DataFrame com os resultados do OHE
#ohe_columns = [f"SaleType_{category}" for category in ohe.categories_[0]]
ohe_df = pd.DataFrame(salestype_ohe, columns=list(ohe.categories_[0]))

# Concatenar o resultado ao DataFrame original
df_categorical = pd.concat([df_categorical, ohe_df], axis=1)

df_categorical.drop(columns=['SaleTypeCat', 'SaleConditionCat'],inplace=True)



In [936]:
# Exibir as primeiras linhas do DataFrame atualizado para verificar o resultado
df_categorical.sample(15)

COD  CWD  Con  ConLD  ConLI  ConLw  New  Oth   WD
756   0.0  0.0  0.0    0.0    0.0    0.0  0.0  0.0  1.0
452   0.0  0.0  0.0    0.0    0.0    0.0  0.0  0.0  1.0
523   0.0  0.0  0.0    0.0    0.0    0.0  1.0  0.0  0.0
770   0.0  0.0  0.0    0.0    0.0    0.0  0.0  0.0  1.0
838   0.0  0.0  0.0    0.0    0.0    0.0  0.0  0.0  1.0
1410  0.0  0.0  0.0    0.0    0.0    0.0  0.0  0.0  1.0
652   0.0  0.0  0.0    0.0    0.0    0.0  0.0  0.0  1.0
55    0.0  0.0  0.0    0.0    0.0    0.0  0.0  0.0  1.0
180   0.0  0.0  0.0    0.0    0.0    0.0  0.0  0.0  1.0
1418  1.0  0.0  0.0    0.0    0.0    0.0  0.0  0.0  0.0
1458  0.0  0.0  0.0    0.0    0.0    0.0  0.0  0.0  1.0
537   1.0  0.0  0.0    0.0    0.0    0.0  0.0  0.0  0.0
530   0.0  0.0  0.0    0.0    0.0    0.0  0.0  0.0  1.0
637   0.0  0.0  0.0    0.0    0.0    0.0  0.0  0.0  1.0
446   0.0  0.0  0.0    0.0    0.0    0.0  0.0  0.0  1.0

## 5- Feature Scaling

**5.1- Apply feature scaling to the variable `GarageArea`. Make sure that the new range fall between `-1/3` and `3`.**

In [937]:
# Apply feature scaling to the variable `GarageArea`. Make sure that the new range fall between `-1/3` and `3`.


print(df_subset['GarageArea'].describe())


count    1460.000000
mean      472.980137
std       213.804841
min         0.000000
25%       334.500000
50%       480.000000
75%       576.000000
max      1418.000000
Name: GarageArea, dtype: float64


In [938]:
# Apply feature scaling to the variable `GarageArea`. Make sure that the new range fall between `-1/3` and `3`.


mmscaler = MinMaxScaler(feature_range=(-1/3, 3))
df_subset['GarageAreaNormalized'] = mmscaler.fit_transform(df_subset[['GarageArea']])


C:\Users\PereiraRodrigues\AppData\Local\Temp\ipykernel_10840\639715783.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_subset['GarageAreaNormalized'] = mmscaler.fit_transform(df_subset[['GarageArea']])


In [939]:
print (f'Valores de garagem normalizados, entre : vamor máximo {df_subset['GarageAreaNormalized'].max()} e valor minimo {df_subset['GarageAreaNormalized'].min()}  ')

Valores de garagem normalizados, entre : vamor máximo 3.0 e valor minimo -0.3333333333333333  


---

## End of Part 1

---

# Part 2

### Part 2 - 50% | Scope

Part 2 entails a project using the same dataset. The goal is to prove your data preprocessing skills. As output from this project, you should delivery the **notebook with the code you have done**.

Please apply, at least, **6 transformations** to the feature set you have in hands (or to the features which makes sense to apply the transformation). Each transformation should be accompanied by an explanation . Last but not least, compare the benefits of such transformation with the baseline score or the last best score. 

Regarding the variables you have to use throughout the Part 2, there are 6 in the total, 3 of them you are free to choose while the remaining 3 I have picked for you:
* YearBuilt
* LotFrontage
* MasVnrType

Make your baseline progressive, i.e. please consider the score from the previous transformation as the new baseline if it shows improvements. Example:

    Baseline - subset of transformations [None]  = 60% accuracy
    Iteration #1 - subset of transformations [A]     = 64% accuracy -> new baseline
    Iteration #2 - subset of transformations [A,B]   = 63% accuracy    (future scaling, ou imputação, ou, pelo menos 6 transformações)
    Iteration #3 - subset of transformations [A,B,C] = 68% accuracy -> new baseline
    ...
    Iteration #N - subset of transformations [A,B,C,..., N]
    (Being A, B, C a transformation that uses 1 or N features.)

Transformation example: encoding `color` & `country` with `One-Hot-Encoding`.

The `target` variable should be used to compute the accuracy (please use the `Target` you have created on the exercise 2.1, part1).


Remark:
* you shall use 6 variable for the assessment
* 3 out of 6 features are designated in the section on the top of the notebook
* the 3 remaining variables are up to you to choose
* you can consider any variable from the original dataset during this assessment

Above all, take this opportunity to practice :)

**Good luck!**

---

In [940]:
df2 = pd.read_csv('data/house_prices_final_project.csv')

Adicionar a lista de colunas parte um, as colunas YearBuilt e MasVncType porque não estavam definidas na lista inicial. 

In [941]:
columns_list.extend(['YearBuilt', 'MasVnrType'])
columns_list

['FullBath',
 'TotRmsAbvGrd',
 'Fireplaces',
 'GarageYrBlt',
 'GarageCars',
 'GarageArea',
 'LotFrontage',
 'WoodDeckSF',
 'OpenPorchSF',
 'SaleType',
 'SaleCondition',
 'SalePrice',
 'YearBuilt',
 'MasVnrType']

In [942]:
# Criar um DataFrame com o subconjunto de colunas necessárias
df_subsetpt2 = df2[columns_list]

In [943]:
df_subset[['YearBuilt', 'MasVnrType']] = df_subsetpt2[['YearBuilt', 'MasVnrType']].copy()



C:\Users\PereiraRodrigues\AppData\Local\Temp\ipykernel_10840\1257106792.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_subset[['YearBuilt', 'MasVnrType']] = df_subsetpt2[['YearBuilt', 'MasVnrType']].copy()
C:\Users\PereiraRodrigues\AppData\Local\Temp\ipykernel_10840\1257106792.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_subset[['YearBuilt', 'MasVnrType']] = df_subsetpt2[['YearBuilt', 'MasVnrType']].copy()


In [944]:
df_subset.sample(10)

FullBath  TotRmsAbvGrd  Fireplaces  GarageYrBlt  GarageCars  GarageArea  \
462          1             5           1       1965.0           1         360   
19           1             6           0       1958.0           1         294   
330          2            10           0       2002.0           1         352   
949          1             5           1       1972.0           2         676   
1435         1             7           1       1962.0           2         462   
39           2             6           0          NaN           0           0   
771          1             6           0       1951.0           2         396   
819          2             6           1       2010.0           2         484   
446          2             6           1       1966.0           2         578   
631          2             6           1       2006.0           2         627   

      LotFrontage  WoodDeckSF  OpenPorchSF SaleType SaleCondition  SalePrice  \
462          60.0           0            0       WD        Normal      62383   
19           70.0           0            0      COD       Abnorml     139000   
330          69.0         155            0       WD        Normal     119000   
949          78.0           0            0       WD        Normal     197500   
1435         80.0           0           36      COD       Abnorml     174000   
39           65.0           0            0       WD       AdjLand      82000   
771          67.0           0            0      COD        Normal     102000   
819          44.0         192           35      New       Partial     224000   
446         137.0           0            0       WD        Normal     190000   
631          34.0         156           73       WD        Normal     209500   

      SalesPriceTarget  LotFrontage_Outlier  GarageAreaNormalized  YearBuilt  \
462                  0                    0              0.512929       1965   
19                   0                    0              0.357781       1958   
330                  0                    0              0.494123       1964   
949                  1                    0              1.255759       1972   
1435                 1                    0              0.752703       1962   
39                   0                    0             -0.333333       1955   
771                  0                    0              0.597555       1951   
819                  1                    0              0.804419       2009   
446                  1                    1              1.025388       1966   
631                  1                    0              1.140574       2006   

     MasVnrType  
462         NaN  
19          NaN  
330     BrkFace  
949         NaN  
1435    BrkFace  
39          NaN  
771         NaN  
819       Stone  
446         NaN  
631       Stone

In [945]:
df_subset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 17 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   FullBath              1460 non-null   int64  
 1   TotRmsAbvGrd          1460 non-null   int64  
 2   Fireplaces            1460 non-null   int64  
 3   GarageYrBlt           1379 non-null   float64
 4   GarageCars            1460 non-null   int64  
 5   GarageArea            1460 non-null   int64  
 6   LotFrontage           1460 non-null   float64
 7   WoodDeckSF            1460 non-null   int64  
 8   OpenPorchSF           1460 non-null   int64  
 9   SaleType              1460 non-null   object 
 10  SaleCondition         1460 non-null   object 
 11  SalePrice             1460 non-null   int64  
 12  SalesPriceTarget      1460 non-null   int64  
 13  LotFrontage_Outlier   1460 non-null   int64  
 14  GarageAreaNormalized  1460 non-null   float64
 15  YearBuilt            

In [946]:
df_subset.describe().T

count           mean           std           min  \
FullBath              1460.0       1.565068      0.550916      0.000000   
TotRmsAbvGrd          1460.0       6.517808      1.625393      2.000000   
Fireplaces            1460.0       0.613014      0.644666      0.000000   
GarageYrBlt           1379.0    1978.506164     24.689725   1900.000000   
GarageCars            1460.0       1.767123      0.747315      0.000000   
GarageArea            1460.0     472.980137    213.804841      0.000000   
LotFrontage           1460.0      69.863699     22.027677     21.000000   
WoodDeckSF            1460.0      94.244521    125.338794      0.000000   
OpenPorchSF           1460.0      46.660274     66.256028      0.000000   
SalePrice             1460.0  180921.195890  79442.502883  34900.000000   
SalesPriceTarget      1460.0       0.501370      0.500169      0.000000   
LotFrontage_Outlier   1460.0       0.060274      0.238075      0.000000   
GarageAreaNormalized  1460.0       0.778515      0.502597     -0.333333   
YearBuilt             1460.0    1971.267808     30.202904   1872.000000   

                                25%            50%            75%       max  
FullBath                   1.000000       2.000000       2.000000       3.0  
TotRmsAbvGrd               5.000000       6.000000       7.000000      14.0  
Fireplaces                 0.000000       1.000000       1.000000       3.0  
GarageYrBlt             1961.000000    1980.000000    2002.000000    2010.0  
GarageCars                 1.000000       2.000000       2.000000       4.0  
GarageArea               334.500000     480.000000     576.000000    1418.0  
LotFrontage               60.000000      69.000000      79.000000     313.0  
WoodDeckSF                 0.000000       0.000000     168.000000     857.0  
OpenPorchSF                0.000000      25.000000      68.000000     547.0  
SalePrice             129975.000000  163000.000000  214000.000000  755000.0  
SalesPriceTarget           0.000000       1.000000       1.000000       1.0  
LotFrontage_Outlier        0.000000       0.000000       0.000000       1.0  
GarageAreaNormalized       0.452985       0.795016       1.020686       3.0  
YearBuilt               1954.000000    1973.000000    2000.000000    2010.0

In [947]:
# Select initial features, including the specified ones
selected_features = ['YearBuilt', 'LotFrontage', 'MasVnrType']
selected_features.extend( ['SaleType', 'Fireplaces', 'GarageCars','SalesPriceTarget'])
df_subsetpt2 = df_subset[selected_features]
df_subsetpt2
#df_subset.sample(5)

YearBuilt  LotFrontage MasVnrType SaleType  Fireplaces  GarageCars  \
0          2003         65.0    BrkFace       WD           0           2   
1          1976         80.0        NaN       WD           1           2   
2          2001         68.0    BrkFace       WD           1           2   
3          1915         60.0        NaN       WD           1           3   
4          2000         84.0    BrkFace       WD           1           3   
...         ...          ...        ...      ...         ...         ...   
1455       1999         62.0        NaN       WD           1           2   
1456       1978         85.0      Stone       WD           2           2   
1457       1941         66.0        NaN       WD           2           1   
1458       1950         68.0        NaN       WD           0           1   
1459       1965         75.0        NaN       WD           0           1   

      SalesPriceTarget  
0                    1  
1                    1  
2                    1  
3                    0  
4                    1  
...                ...  
1455                 1  
1456                 1  
1457                 1  
1458                 0  
1459                 0  

[1460 rows x 7 columns]

In [948]:
df_subsetpt2.isnull().mean() * 100

YearBuilt            0.000000
LotFrontage          0.000000
MasVnrType          59.726027
SaleType             0.000000
Fireplaces           0.000000
GarageCars           0.000000
SalesPriceTarget     0.000000
dtype: float64

In [949]:
# Baseline score
#train_df, test_df = split_dataset(df_subsetpt2, 'SalesPriceTarget')
#baseline_score = score_approach(train_df, test_df, 'SalesPriceTarget')
#print(f"Baseline score: {baseline_score}")

In [950]:
#Tratamento  MasVnrType

df_subsetpt2["MasVnrType"] = pd.Categorical(df_subsetpt2["MasVnrType"],ordered=True,
    categories=["None", "CBlock", "BrkCmn", "BrkFace", "Stone"]
)

C:\Users\PereiraRodrigues\AppData\Local\Temp\ipykernel_10840\3522395156.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_subsetpt2["MasVnrType"] = pd.Categorical(df_subsetpt2["MasVnrType"],ordered=True,


In [951]:
#print(df_subsetpt2["MasVnrType"].cat.categories)
print(df_subsetpt2["MasVnrType"].head(10))

0    BrkFace
1        NaN
2    BrkFace
3        NaN
4    BrkFace
5        NaN
6      Stone
7      Stone
8        NaN
9        NaN
Name: MasVnrType, dtype: category
Categories (5, object): ['None' < 'CBlock' < 'BrkCmn' < 'BrkFace' < 'Stone']


In [952]:
# Substituir os valores em falta pela moda
df_subsetpt2.groupby('MasVnrType')['MasVnrType'].count()

C:\Users\PereiraRodrigues\AppData\Local\Temp\ipykernel_10840\1471370338.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_subsetpt2.groupby('MasVnrType')['MasVnrType'].count()


MasVnrType
None         0
CBlock       0
BrkCmn      15
BrkFace    445
Stone      128
Name: MasVnrType, dtype: int64

In [953]:
imputer = SimpleImputer(strategy='most_frequent')
df_subsetpt2[['MasVnrType']] = imputer.fit_transform(df_subsetpt2[['MasVnrType']])


C:\Users\PereiraRodrigues\AppData\Local\Temp\ipykernel_10840\2943377578.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_subsetpt2[['MasVnrType']] = imputer.fit_transform(df_subsetpt2[['MasVnrType']])


In [954]:
# Substituir os valores em falta pela moda
df_subsetpt2.groupby('MasVnrType')['MasVnrType'].count()

MasVnrType
BrkCmn       15
BrkFace    1317
Stone       128
Name: MasVnrType, dtype: int64